### PR TITLE
cli: Customize max sign attempts for deploy and write-buffer

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

People are still having issues getting their program deployments to land.

#### Summary of Changes

As a simple extra knob for users, allow them to specify the maximum number of signing attempts.

While updating the changelog, I noticed that `--with-compute-unit-price` wasn't documented, so I added that in the same line

### CODE REVIEW

**Comment by jstarry** [comment]
Change looks good but I don't really understand what a user is supposed to set this to.. if they are having problems is the idea that they should add more retries? Couldn't they just restart the deploy?

**Comment by joncinque** [comment]
> I don't really understand what a user is supposed to set this to..

That's a good point, maybe the help text can be clearer to help with that? It defaults to 5, so how about we say "Transactions are resent as long as the blockhash is still valid, which is a minimum of 60 seconds. For the default value of 5, the tool will sign transactions a maximum of 5 times, which means it will keep sending transactions for at least 5 minutes or until all transactions are confirmed, whichever comes first."

> if they are having problems is the idea that they should add more retries? Couldn't they just restart the deploy?

Exactly. when the network is congested and they're having a hard time landing transactions, users can set a huge number and just walk away without having to constantly kick it off again by hand.

Your previous change to read the account before creating the transaction makes it easier to restart the deploy, but it still requires someone to manually restart the deployment.

**Comment by jstarry** [comment]
Yeah the extra context seems quite helpful!

I think "resent" is a bit confusing since it's not necessarily known that the deploy process is continuously resending transactions.

Here's my attempt, though I'm happy with yours too!

If any transactions sent during the program deploy are still unconfirmed after the initially chosen recent blockhash expires, those transactions will be resigned with a new recent blockhash and resent. Use this setting to adjust the maximum number of transaction signing iterations. Each blockhash is valid for about 60 seconds, which means using the default value of 5 will lead to sending transactions for at least 5 minutes or until all transactions are confirmed, whichever comes first

**Comment by joncinque** [comment]
I went with yours, thanks for the suggestion :grinning:

File: cli/src/program.rs [hunk 1]
```diff
@@ -246,6 +248,26 @@ impl ProgramSubCommands for App<'_, '_> {
...
+                                    If any transactions sent during the program deploy are still \
+                                    unconfirmed after the initially chosen recent blockhash \
+                                    expires, those transactions will be resigned with a new \
+                                    recent blockhash and resent. Use this setting to adjust \
+                                    the maximum number of transaction signing iterations. Each \
+                                    blockhash is valid for about 60 seconds, which means using \
+                                    the default value of 5 will lead to sending transactions \
+                                    for at least 5 minutes or until all transactions are confirmed,\
+                                    whichever comes first.",
+                                ),
```
Comment (jstarry) [inline cli/src/program.rs]: The line break is a little funky

```
        --max-sign-attempts <max_sign_attempts>
            Maximum number of attempts to sign or resign transactions after blockhash expiration.
                                                If any transactions sent during the program deploy are still unconfirmed
            after the initially chosen recent blockhash expires, those transactions will be resigned with a new recent
            blockhash and resent. Use this setting to adjust the maximum number of transaction signing iterations. Each
            blockhash is valid for about 60 seconds, which means using the default value of 5 will lead to sending
            transactions for at least 5 minutes or until all transactions are confirmed,whichever comes first. [default:
            5]
```

Comment (joncinque) [reply cli/src/program.rs]: Ah oops, sorry about that. Thanks for noticing -- fixed!

**Review by jstarry — APPROVED** [review]

**Pull request merged**